### ISSUE TITLE
Intersection between a quadratic bezier segment and a line is not found

### ISSUE THREAD

**Issue by Dollab** [nical/lyon#678]
Hi,
I've come across a case where the intersection between a quadratic and a line fails to return the proper value. Probably a floating point issue again. I think it has to do with the way the quadratic is converted to a cubic.

Here is the failing exemple:
```rust
#[test]
fn test_quad_intersection_with_line() {
        let points = [
            [-7768.80859375, -35563.80859375],
            [-38463.125, -10941.41796875],
            [-21846.12890625, -13518.1953125],
            [-11727.439453125, -22080.33203125],
        ];

        let quadratic = QuadraticBezierSegment {
            from: Point::new(points[0][0], points[0][1]),
            ctrl: Point::new(points[1][0], points[1][1]),
            to: Point::new(points[2][0], points[2][1]),
        };

        let line = Line {
            point: Point::new(points[3][0], points[3][1]),
            vector: Vector::new(-0.5, -0.5).normalize(),
        };

        let intersections = quadratic.line_intersections(&line);
        assert_eq!(intersections.len(), 1, "We should find an intersection point");
}
```
One way to solve this would be to implement the quadratic intersection directly (without the conversion to cubic).

![image](https://user-images.githubusercontent.com/7534544/123938197-2fca2c80-d997-11eb-990c-df00f98b3a80.png)

**Comment by nical** [comment]
Hi, thanks for filing this.

The problem is indeed caused by float precision, Running your test as is actually passes because the scalars will be `f64`, but forcing them to be `f32` shows the issue.
The conversion from to cubic isn't a problem here. It'd be nice to have this implemented for quadratic curves directly for the sake of performance but corresponding cubic curve is adequately represented, yet it fails to find the intersection.

The issue happens because the intersection code uses constant epsilon values to determine if some quantities are too close to zero in which case precision drops and a different equation is used. For this to really work we would need very small epsilons when parameters are small and larger epsilons when parameters are large. The epsilon is apparently too small for this test case with `f32` precision.

I'll try to come up with a good way to pick epsilons in function of magnitudes and the scalar precision, in the mean time my advice for you is to default to `f64` for intersection checks (regardless of how you store the values in the rest of your code). I do that in the fill tessellator for line segment intersections and convert the results back to `f32` immediately.

**Comment by Dollab** [comment]
Hi, thanks a lot for the fast reply and for having looked at the problem. I will switch to f64 as suggested. 
Thanks again !

**Comment by nical** [comment]
I published a fix for the f32 precision in lyon_geom 0.17.4 so it should be less awful although I still recommend using f64 for this type of thing (don't hesitate to continue reporting other precision issues eitherway!).
